feat: add support for string-based custom queries

paullewis opened the pull request:
Relevant issues: #5747, #5405, #4171.

paullewis committed:
feature: Add support for string-based custom queries

mathiasbynens commented on src/JSHandle.ts:
```diff
@@ -426,8 +429,18 @@ export class ElementHandle extends JSHandle {
-        (element, selector) => element.querySelector(selector),
+    let handler = (element: Element, selector: string): ElementHandle => element.querySelector(selector) as unknown as ElementHandle;
+    if (selector.indexOf(SEARCH_SEPARATOR) !== -1) {
+      const [name, updatedSelector] = selector.split(SEARCH_SEPARATOR);
```
If you're gonna use this to create a custom XPath-based query handler, you'll probably use `/` quite often in your selectors. With this implementation, `updatedSelector` only contains everything between the first `/` and the second `/`. We could instead use the result of `indexOf` with `slice` to get the entire selector.

mokeyish replied:
> If you're gonna use this to create a custom XPath-based query handler, you'll probably use `/` quite often in your selectors. With this implementation, `updatedSelector` only contains everything between the first `/` and the second `/`. We could instead use the result of `indexOf` with `slice` to get the entire selector.

What about this?
<img width="475" alt="图片" src="https://user-images.githubusercontent.com/16131917/80323134-193cb880-885c-11ea-9060-cbb85b6cd541.png">

mathiasbynens replied:
With this patch, you could implement everything in that table yourself, without being limited to what's in the table!

paullewis replied:
Done.

mathiasbynens commented on src/Puppeteer.js:
```diff
@@ -144,4 +146,23 @@ module.exports = class {
+   * @param {string} name
+   * @param {!Function} queryFunction
+   */
+  registerCustomQueryFunction(name, queryFunction) {
```
bikeshed: `Function` sounds a little generic. How about `Handler`?

paullewis replied:
I was aware that this file was likely going over to TS at some point soon, and therefore would get a "proper" signature at that point. AFAICT you can't easily specify JSDoc for function params, but maybe I missed something.

jackfranklin replied:
👍 on this - let's leave this and be much more explicit and define nicer types when we migrate this file to TS

mathiasbynens replied:
I was talking about the method name (`registerCustomQueryFunction`), not the types

paullewis replied:
Well that's changed now, too! 🥳

mathiasbynens commented on test/elementhandle.spec.js:
```diff
@@ -248,4 +248,25 @@ describe('ElementHandle specs', function() {
+      await page.setContent('<div id="not-foo"></div><div id="foo"></div>');
+
+      // Register.
+      puppeteer.registerCustomQueryFunction('get-by-id', (element, selector) => document.querySelector(`[id="${selector}"]`));
```
I think it'd be nice to limit query handler names to `/^[a-zA-Z]+$/`, matching built-in JavaScript APIs. We could validate this upon registration. WDYT? In this case, it'd be `getById` instead of `get-by-id`.

mokeyish replied:
I think  using `id:elementId` will be better than `get-by-id/elementId`. 

<img width="475" alt="图片" src="https://user-images.githubusercontent.com/16131917/80323134-193cb880-885c-11ea-9060-cbb85b6cd541.png">

mathiasbynens replied:
When registering your own query handler you could name it `id` and then it would be `id/elementId`, which seems close enough?

The reason we cannot use `:` is because it has special meaning in CSS, so selectors like `p:first` become ambiguous.

mathiasbynens commented on src/QueryFunction.ts:
```diff
@@ -0,0 +1,47 @@
+const _customQueryFunctions = new Map<string, QueryFunction>();
+
+export function registerCustomQueryFunction(name: string, queryFunction: Function): void {
+  if (_customQueryFunctions.get(name))
```
TODO: validate the name here. At a minimum, we should ensure it doesn't contain `SEARCH_SEPARATOR`. But IMHO we should be even more restrictive and only allow `/^[a-zA-Z]+$/`.

jackfranklin replied:
👍 and then let's add some unit tests to check the behaviour

paullewis replied:
Sure

paullewis replied:
Added validation, but our current setup isn't geared up for unit tests particularly. I have a test that asserts it throws with invalid names.

mathiasbynens commented on src/JSHandle.ts:
```diff
@@ -442,10 +455,18 @@ export class ElementHandle extends JSHandle {
-    );
+    let handler = (element: Element, selector: string): ElementHandle[] => element.querySelectorAll(selector) as unknown as ElementHandle[];
+    if (selector.indexOf(SEARCH_SEPARATOR) !== -1) {
+      const [name, updatedSelector] = selector.split(SEARCH_SEPARATOR);
```
Same here.

mathiasbynens commented on src/JSHandle.ts:
```diff
@@ -442,10 +455,18 @@ export class ElementHandle extends JSHandle {
-        selector
-    );
+    let handler = (element: Element, selector: string): ElementHandle[] => element.querySelectorAll(selector) as unknown as ElementHandle[];
+    if (selector.indexOf(SEARCH_SEPARATOR) !== -1) {
```
Can we extract this common `indexOf` + `split` logic?

paullewis replied:
Done.

mathiasbynens replied:
Instead of just `_getQueryFunctionNameAndSelector`, could we extract the entire block? It's shared between `$$` and `$$eval` below.

```js
    let handler = (element: Element, selector: string): ElementHandle => element.querySelector(selector) as unknown as ElementHandle;
    if (selector.indexOf(SEARCH_SEPARATOR) !== -1) {
      const {name, updatedSelector} = this._getQueryFunctionNameAndSelector(selector);
      const queryFunction = customQueryHandlers().get(name);
      if (!queryFunction)
        throw new Error(`$ query set to use "${name}", but no query function of that name was found`);

      handler = queryFunction as typeof handler;
      selector = updatedSelector;
    }
```

Could be something like:

```js
const {handler, selector} = getHandlerAndSelector();
```

(with possibly better names)

WDYT?

mathiasbynens commented on test/elementhandle.spec.js:
```diff
@@ -248,4 +248,25 @@ describe('ElementHandle specs', function() {
+      try {
+        await page.$('get-by-id/foo');
+        expect.fail('Custom query function not set - throw expected');
+      } catch (e) {
```
nit: `error`

jackfranklin replied:
we could consider adding this eslint plugin/rule: https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/rules/catch-error-name.md to enforce this

mathiasbynens replied:
let's do it!

jackfranklin replied:
Will do in a follow up PR :)

mathiasbynens reviewed (changes requested):
Looking great so far! Left some thoughts, PTAL.

jackfranklin commented on src/QueryFunction.ts:
```diff
@@ -0,0 +1,47 @@
+  (element: Element, selector: string): (ElementHandle | null) | ElementHandle[];
+}
+
+const _customQueryFunctions = new Map<string, QueryFunction>();
```
nit: could we lose the `_` here? I think the fact it's not exported is enough.

paullewis replied:
It chucks a lint error: `Individual declarations in merged declaration 'customQueryFunctions' must be all exported or all local.`

paullewis replied:
It throws a lint error :(

paullewis committed:
Addresses feedback; adds more tests

paullewis committed:
Validates name

paullewis commented:
I _think_ that's all the feedback addressed. PTAL.

(@mathiasbynens I will namespace this under experimental as discussed when everything else is okay)

paullewis committed:
Changes function to handler; adds experimental

mathiasbynens commented on src/QueryHandler.ts:
```diff
@@ -0,0 +1,51 @@
+  if (_customQueryHandlers.get(name))
+    throw new Error(`A custom query function named "${name}" already exists`);
+
+  const isValidName = /^[a-zA-Z\-]+$/.test(name);
```
nit: no need to escape `-` as last char in character class

```suggestion
  const isValidName = _/^[a-zA-Z-]+$/_.test(name);
```

but as I said earlier I'd prefer simplifying this to `[a-zA-Z]`, so that query name handlers are (roughly) similar to JS API function names

paullewis replied:
Done

mathiasbynens commented on src/JSHandle.ts:
```diff
@@ -467,10 +497,18 @@ export class ElementHandle extends JSHandle {
+      const {name, updatedSelector} = this._getQueryFunctionNameAndSelector(selector);
+      const queryFunction = customQueryHandlers().get(name);
+      if (!queryFunction)
+        throw new Error(`$$ query set to use "${name}", but no query function of that name was found`);
```
`$$` -> `$$eval`, but we don't need to be this specific here. If we abstract this into a shared helper between `$$eval` and `$$` as suggested above this will automatically be resolved

mathiasbynens commented on src/Puppeteer.js:
```diff
@@ -34,6 +35,7 @@ module.exports = class {
     this._isPuppeteerCore = isPuppeteerCore;
     // track changes to Launcher configuration via options or environment variables
     this.__productName = productName;
+    this._customQueryFunctions = new Map();
```
One last thing, let's do a global rename for "queryFunction" to "queryHandler" so that the naming is internally consistent

paullewis replied:
Actually, that Map isn't necessary. Removed.

paullewis committed:
Moves code to helper

mathiasbynens commented on src/QueryHandler.ts:
```diff
@@ -0,0 +1,51 @@
+
+const _customQueryHandlers = new Map<string, QueryHandler>();
+
+export function registerCustomQueryHandler(name: string, queryFunction: Function): void {
```
```suggestion
export function registerCustomQueryHandler(name: string, handler: Function): void {
```

...and similar below

paullewis committed:
Renames to handler

paullewis committed:
Updates error messages

mathiasbynens commented on src/JSHandle.ts:
```diff
@@ -442,10 +464,11 @@ export class ElementHandle extends JSHandle {
-    );
+    const defaultHandler = (element: Element, selector: string): ElementHandle[] => element.querySelectorAll(selector) as unknown as ElementHandle[];
+    const {updatedSelector, queryHandler} = this._getQueryHandlerAndSelector(selector, defaultHandler,
+        '$$ query set to use "QUERY_HANDLER_NAME", but no query function of that name was found');
```
Looks like the error message is always the same. We could either parameterize the function name (`$`, `$$`, or `$$eval`) or just omit it. I think omitting seems fine:

    'Query set to use "QUERY_HANDLER_NAME", but no query function of that name was found'

That way we no longer need the `errorMessage` parameter.

paullewis replied:
Done

paullewis committed:
Deparameterize fn

mathiasbynens commented on test/elementhandle.spec.js:
```diff
@@ -248,4 +248,53 @@ describe('ElementHandle specs', function() {
+      const {page, puppeteer} = getTestState();
+      await page.setContent('<div id="not-foo"></div><div class="foo">Foo1</div><div class="foo baz">Foo2</div>');
+      puppeteer.__experimental_registerCustomQueryHandler('getbyclass', (element, selector) => document.querySelectorAll(`.${selector}`));
+      const elements = await page.$$eval('getbyclass/foo', divs => divs.length);
```
nit: let's make these `getByClass` etc. to match built-in JS API naming

paullewis committed:
Renames query handlers

mathiasbynens commented on src/JSHandle.ts:
```diff
@@ -425,11 +428,29 @@ export class ElementHandle extends JSHandle {
+    const updatedSelector = selector.slice(index + SEARCH_SEPARATOR.length);
+    const queryHandler = customQueryHandlers().get(name);
+    if (!queryHandler)
+      throw new Error('Query set to use "QUERY_HANDLER_NAME", but no query handler of that name was found'.replace('QUERY_HANDLER_NAME', name));
```
```suggestion
      throw new Error(`Query set to use "${name}", but no query handler of that name was found`);
```

paullewis committed:
Revert back to Function

paullewis committed:
Fix expectation

paullewis commented:
Okay, these are the call sites. I'll go through and confirm that each one lands in the correct place to have custom query support.

 - [X] page.$(selector)
 - [X] page.$$(selector)
 - [X] page.$$eval(selector, pageFunction[, ...args])
 - [X] page.$eval(selector, pageFunction[, ...args])
 - [X] page.click(selector[, options])
 - [X] page.focus(selector)
 - [X] page.hover(selector)
 - [X] page.select(selector, ...values)
 - [X] page.tap(selector)
 - [X] page.type(selector, text[, options])
 - [X] page.waitFor(selectorOrFunctionOrTimeout[, options[, ...args]])
 - [X] page.waitForSelector(selector[, options])
 - [X] frame.$(selector)
 - [X] frame.$$(selector)
 - [X] frame.$$eval(selector, pageFunction[, ...args])
 - [X] frame.$eval(selector, pageFunction[, ...args])
 - [X] frame.click(selector[, options])
 - [X] frame.focus(selector)
 - [X] frame.hover(selector)
 - [X] frame.select(selector, ...values)
 - [X] frame.tap(selector)
 - [X] frame.type(selector, text[, options])
 - [X] frame.waitFor(selectorOrFunctionOrTimeout[, options[, ...args]])
 - [X] frame.waitForSelector(selector[, options])
 - [X] elementHandle.$(selector)
 - [X] elementHandle.$$(selector)
 - [X] elementHandle.$$eval(selector, pageFunction[, ...args])
 - [X] elementHandle.$eval(selector, pageFunction[, ...args])

paullewis committed:
Adds support for waitFor{Selector}

paullewis committed:
Fixes merge issues

paullewis commented:
Had to force push because of all sorts of conflicts that I presumably made worse during rebase :(

Anyway, PTAL

paullewis committed:
Changes to undefined

paullewis committed:
Removes unnecessary const

jackfranklin committed:
chore: migrate src/Target to TypeScript (#5771)

mathiasbynens commented on src/DOMWorld.ts:
```diff
@@ -430,18 +436,24 @@ class WaitTask {
     this._polling = polling;
     this._timeout = timeout;
-    this._predicateBody = helper.isString(predicateBody) ? 'return (' + predicateBody + ')' : 'return (' + predicateBody + ')(...args)';
+    this._predicateBody = helper.isString(predicateBody) ? 'return (' + predicateBody + ')' : wrapper;
```
This is getting complicated enough to move to its own helper like

```suggestion
    this._predicateBody = getPredicateBody(predicateBody, predicateQueryHandlerBody);
```

which then does

```js
if (helper.isString(predicateBody)
  return `return (${predicateBody});`;
if (predicateQueryHandlerBody) {
  return `
    return (function wrapper(args) {
      const predicateQueryHandler = ${predicateQueryHandlerBody};
      return (${predicateBody})(...args);
    })(args);`;
}
return `return (${predicateBody})(...args);`;
```

mathiasbynens commented on src/QueryHandler.ts:
```diff
@@ -0,0 +1,74 @@
+  const updatedSelector = selector.slice(index + 1);
+  const queryHandler = customQueryHandlers().get(name);
+  if (!queryHandler)
+    throw new Error('Query set to use "QUERY_HANDLER_NAME", but no query handler of that name was found'.replace('QUERY_HANDLER_NAME', name));
```
```suggestion
    throw new Error(`Query set to use "${name}", but no query handler of that name was found`);
```

mathiasbynens reviewed (approved):
LGTM! Super excited about this :fire::fire::fire:

Left some suggestions for cleanup/polish, but I'm fine with doing those as follow-up changes in a separate PR if you prefer.

paullewis committed:
Update src/QueryHandler.ts

Co-Authored-By: Mathias Bynens <mathias@qiwi.be>

paullewis committed:
Update src/DOMWorld.ts

Co-Authored-By: Mathias Bynens <mathias@qiwi.be>

paullewis committed:
Fixes lint issues

mlrv committed:
chore: update incorrect link for DeviceDescriptors (#5777)

mathiasbynens committed:
chore: disable flaky setUserAgent test in Firefox (#5780)

jackfranklin committed:
chore: migrate src/NetworkManager to TypeScript (#5774)

jackfranklin committed:
docs(contributing): update per recent changes (#5778)

Co-authored-by: Mathias Bynens <mathias@qiwi.be>

jackfranklin committed:
chore: enable mocha retries (#5782)

We are seeing some CI flakes due to network resets, or issues out of our control. Let's try to avoid some of those by using Mocha's built-in retries.

mathiasbynens committed:
chore: fix doclint issues (#5784)

paullewis commented:
Changed CLA to yes because all commits were by people who have signed the CLA.

jackfranklin committed:
chore: log product + binary on unit test runs (#5785)

This way it's super clear which product and binary you're testing against.

paullewis commented:
Changed CLA to yes (again) because all commits were by people who have signed the CLA.
